File: Hammerspoon Tests/HScoresetupHelper.m
#import "HScoresetupHelper.h"
#import "MJLua.h"

static BOOL testFlag = NO;

static int verifyShutdown(lua_State *L) {
    testFlag = YES;
    return 0;
}

@implementation HScoresetupHelper

+ (void)registerShutdownLib {
    luaL_Reg shutdownLib[] = {
        {"verifyShutdown", verifyShutdown},
        {NULL, NULL}
    };
    LuaSkin *skin = [LuaSkin sharedWithState:NULL];
    [skin registerLibrary:"shutdownLib" functions:shutdownLib metaFunctions:nil];
    lua_setglobal(skin.L, "shutdownLib");
}

+ (BOOL)shutdownFired {
    return testFlag;
}

+ (void)resetShutdownFlag {
    testFlag = NO;
}

@end

File: Packages/HammerspoonTests/HScoresetupHelper.swift
import LuaSkin

nonisolated(unsafe) private var testFlag = false

private let verifyShutdown: lua_CFunction = { L in
    testFlag = true
    return 0
}

enum HScoresetupHelper {
    @MainActor static func registerShutdownLib() {
        var shutdownLib: [luaL_Reg] = [
            luaL_Reg(name: strdup("verifyShutdown"), func: verifyShutdown),
            luaL_Reg(name: nil, func: nil),
        ]
        let skin = LuaSkin.shared(with: nil) as! LuaSkin
        skin.registerLibrary("shutdownLib", functions: &shutdownLib, metaFunctions: nil)
        lua_setglobal(skin.l, "shutdownLib")
    }

    static func shutdownFired() -> Bool { testFlag }
    static func resetShutdownFlag() { testFlag = false }
}
